agent/secrets.go

```diff
@@ -10,44 +10,75 @@ import (
 // mapped by secret ID
 type secrets struct {
 	mu sync.RWMutex
-	m  map[string]api.Secret
+	m  map[string]*api.Secret
 }
 
 func newSecrets() *secrets {
 	return &secrets{
-		m: make(map[string]api.Secret),
+		m: make(map[string]*api.Secret),
 	}
 }
 
-// Get returns a secret by ID.  If the secret doesn't exist, returns nil.
-func (s *secrets) Get(secretID string) api.Secret {
+// get returns a secret by ID.  If the secret doesn't exist, returns nil.
+func (s *secrets) get(secretID string) *api.Secret {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	return s.m[secretID]
+	if s, ok := s.m[secretID]; ok {
+		return s
+	}
+	return nil
 }
 
-// Add adds one or more secrets to the secret map
-func (s *secrets) Add(secrets ...api.Secret) {
+// add adds one or more secrets to the secret map
+func (s *secrets) add(secrets ...api.Secret) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	for _, secret := range secrets {
-		s.m[secret.ID] = secret
+		s.m[secret.ID] = &secret
```

Comment:
This looks wrong. You'll add the address of the iterator to the map multiple times.

`s.m[secrets.ID] = secret.Copy()`?

Or potentially make `add` take `secrets ...*api.Secret` to avoid copies, if that makes sense.